cmd/rpc/client.go

```diff
@@ -60,29 +61,37 @@ func (client *Client) Call(serviceMethod string, args, reply interface{}) error
 
 	var buf bytes.Buffer
 	if err = gob.NewEncoder(&buf).Encode(callRequest); err != nil {
-		return err
+		return nil, err
 	}
 
-	response, err := client.httpClient.Post(client.serviceURL.String(), "", &buf)
+	var body io.Reader = &buf
+	if reader != nil {
+		body = io.MultiReader(body, reader)
+	}
+
+	response, err := client.httpClient.Post(client.serviceURL.String(), "", body)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer response.Body.Close()
 
 	if response.StatusCode != http.StatusOK {
-		return fmt.Errorf("%v rpc call failed with error code %v", serviceMethod, response.StatusCode)
+		response.Body.Close()
+		return nil, fmt.Errorf("%v rpc call failed with error code %v", serviceMethod, response.StatusCode)
 	}
 
 	var callResponse CallResponse
 	if err := gob.NewDecoder(response.Body).Decode(&callResponse); err != nil {
-		return err
+		response.Body.Close()
+		return nil, err
 	}
 
 	if callResponse.Error != "" {
-		return errors.New(callResponse.Error)
+		response.Body.Close()
+		return nil, errors.New(callResponse.Error)
 	}
 
-	return gobDecode(callResponse.ReplyBytes, reply)
+	return response.Body, gobDecode(callResponse.ReplyBytes, reply)
```

Comment:
Unable to understand the intention of returning response.Body here. Because code is doing `defer response.Body.Close()` even otherwise, gob-decoder would have read all the contents from response.Body

Reply:
`defer response.Body.Close()` must be removed.  `gobDecode()` or `gob.NewDecoder(bytes.NewReader(data)).Decode(e)` does not read entire stream.